bin/sort_includes.py

```diff
@@ -0,0 +1,125 @@
+
+def sorted_includes(block):
+    """
+    Sorts the include statements in `block`.
+
+    :param block: source code chunk containing 1 or more include statements
+    :return: `block` with the include statements sorted and a blank line
+             between user and sys includes
+    """
+
+    # Replace blank lines with an include string that sorts after user
+    # includes but before sys includes
+    lines = block.splitlines()
+
+    user_includes = [line for line in lines if '"' in line]
+    sys_includes = [line for line in lines if '<' in line]
+    blank_lines = [line for line in lines if line == ""]
+
+    assert len(lines) == len(user_includes) + len(sys_includes) + len(blank_lines)
+
+    user_includes.sort(key=lambda line: line[line.find('"'):])
+    sys_includes.sort(key=lambda line: line[line.find('<'):])
+
+    if user_includes and sys_includes and not blank_lines:
+        blank_lines = [""]
+
+    # Sort lines and undo blank line replacement
+    lines = user_includes + blank_lines + sys_includes
+
+    # Join sorted lines back into a block
+    return "\n".join(lines) + "\n"
+
+def sort_includes(path, args):
+    """
+    Processes the C++ source file in `path` to sort its include statements.
+
+    :param path: a Path object for a C++ source file
+    :param args: command line configuration
+    :return: True if sorting changes were made, False otherwise
+    """
+    source = path.read_text()
+    new_source = ""
+
+    end = 0
+    for m in includes_re.finditer(source):
+        if m.start() != end:
+            new_source += source[end:m.start()]
+        new_source += sorted_includes(m.group())
+        end = m.end()
+
+    if not end:
+        # No includes found
+        return False
+
+    new_source += source[end:]
+    if new_source != source:
+        if args.update:
+            path.write_text(new_source)
+        print(f"Includes in {path} were unsorted")
+        return True
+    return False
+
+if __name__ == "__main__":
+    desc = """
+    Processes C++ source files to check if their include statements are sorted alphabetically.
+    Include statements with any non-space characters after the closing `"` or `>` will not
+    be re-ordered.
+    The exit value is the number of files that had unsorted include statements.
+    """
+    p = argparse.ArgumentParser(description=desc)
+    p.add_argument("--update", action="store_true", help="update files to sort includes")
+    p.add_argument("path", help="C++ file or directory containing C++ files", nargs="+")
+
+    args = p.parse_args()
+
+    unsorted = 0
+    files = []
+    for name in args.path:
+        arg = Path(name)
+        if arg.is_file():
+            files.append(arg)
+        else:
+            for dirpath, dirnames, filenames in os.walk(arg):
+                for filename in filenames:
+                    file = Path(dirpath).joinpath(filename)
+                    if file.suffix in (".cpp", "hpp"):
```

Comment:
`"hpp"` -> `".hpp"`

This bug explains why there are no modified `*.hpp` files in the PR. And I've discovered that blindly sorting includes in these files (especially `*.inline.hpp` files) causes building to break quickly. This is another reason that the more incremental approach I suggest at https://github.com/openjdk/jdk/pull/24180#issuecomment-2752572125

Reply:
I was going to ask that, but forgot about it once I had looked through all the files. :)